### PR TITLE
Refactor plugins using Delete2 & Chgrp2 requests

### PR DESCRIPTION
**Pull request by ximenesuk**
This PR refactors the CLI plugins `delete` and `chgrp` to use the new `Delete2` and `Chgrp2` requests respectively. In addition `fs rename` has been modified to use `Delete2`.

The `--edit` option has been removed as it was broken by the earlier changes

The `--opt` option have been removed and its usage replaced by two new options `--include` and `--exclude`. How the old option maps to the new ones is detailed in the testing section below.

A new option `--ordered` has been introduced to force the preservation of the order of the arguments.

Integration tests have been added but, as ever, more would be useful to test more complex functionality.

Testing
-------
Relevant cli unit and integration tests should pass.

A number of simple delete and chgrp should be checked, see `--help` for examples. Also see below for how the new usage maps to the old usage. In addition more complex "skiphead" requests should be tried, for examples see below.

Prior to this PR objects had to be listed separately. So to delete three images and two datasets.
```
$ omero delete /Image:51 /Image:52 /Image:53 /Dataset:101 /Dataset:201
```
While this form can still be used the following more concise version is now possible.
```
$ omero delete Image:51,52,53 Dataset:101,201
```
Note also that the preceding `/` is no longer used though the command should still work if it is.

Prior to this PR including or excluding objects from a delete used the `KEEP` and `HARD` syntax with `--opt`. So to delete and image but keep its annotations:
```
$ omero delete --opt /Annotation:KEEP /Image:51 
```
This form is no longer possible. Instead the `--exclude` option should be used.
```
$ omero delete Image:51 --exclude Annotation
```
ie the annotations are *excluded* from the delete. Multiple classes of object can be excluded:
```
$ omero delete Image:51,52,53 --exclude TagAnnotation,FileAnnotation
```
Similarly uses of `/Object:HARD` would be replaced by `--include Object`. 

All of this also applies to `chgrp` so to change the group of dataset but not its images.
```
$ omero chgrp 101 Dataset:201 --exclude Image
```
This may orphan the images if they are not also in another dataset.

To delete or change the group of all the objects under some other object a `SkipHead` command is used behind the scenes. CLI users don't need to worry about the details this but the CLI syntax is important.

Delete all images under a project an object hierarchy is used. 
```
$ omero delete Project/Dataset/Image:53
```
As only the first and last objects are important this hierarchy can be shortened
```
$ omero delete Project/Image:53
```
Both of these commands will delete **all** of the images under Project 53.

Links can also be deleted, so to effectively "orphan" all images under projects 51 and 52 and also those under dataset 53:
```
$ omero delete Project/Dataset/DatasetImageLink:51,52 Dataset/DatasetImageLink:53
```
With all of the above commands the option `--ordered` can be used. Without the option,
```
$ omero delete Image:51 Image:52 Image:53 Dataset:101 Dataset:201
```
will be rolled into one single delete command that is passed to the server while
```
$ omero delete Image:51 Image:52 Image:53 Dataset:101 Dataset:201 --ordered
```
will be processed as five separate commands in the order given on the command line. The effect here is that there should be no difference between the two calls but the `--ordered` option is added for any potential edge cases where the order may be important.

One very specific case to test is that of a MIF. A two image MIF with images 51 and 52 should have the following results according to several different calls:
```
$ omero delete Image:51                    # failure, one image cannot be deleted (no change to behaviour here)
$ omero delete Image:51,52                 # success
$ omero delete Image:51 Image:52           # success
$ omero delete Image:51 Image:52 --ordered # failure, as there are two commands they cannot succeed separately
```

Finally a couple of smaller changes need to be checked, the commands should work as before.

The subcommand `group perms` should be checked:
```
$ omero group perms --name pg-0 --perms "rwr---"
Changed permissions for group pg-0 (id=3) to rwr---
$ omero group perms --id 3 --perms "rw----"
Changed permissions for group pg-0 (id=3) to rw----
```

And `omero fs rename` should be checked. See https://github.com/openmicroscopy/openmicroscopy/pull/2805#issue-37880873

--no-rebase

### CODE REVIEW

**Comment by joshmoore** [comment]
NB and/or RFE: how difficult would it be to integrate the `dry-run` logic at this point?

**Comment by ximenesuk** [comment]
It's up there... https://github.com/ximenesuk/openmicroscopy/commit/2b32d37e60e0fd11a64986578fe8868fea086316

or do you mean more than that?

**Comment by ximenesuk** [comment]
Also, for an RFE, at the moment something like:
```
omero delete Project/Dataset/DatasetImageLink:51,52 Dataset/DatasetImageLink:53
```
would be resolved into two commands wrapped in a `DoAll`, but `Delete2` should allow this to be a single command. That would either involve changing the way `GraphArg` works or, more simply, parsing and combining the commands before submitting a single request. The second option would be relatively straightforward I think. Whether it buys us any improved efficiency I don't know. /cc @mtbc

**Comment by mtbc** [comment]
It may actually be more useful for chgrp in allowing unexpectedly linked indirectly referenced things to be moved while remaining linked to all of the graphs that contain them. It could theoretically make delete more efficient too, but I don't know that that's where the real win would be.

Actually separate ordered delete requests can be useful for things like deletion of original file hierarchies, but the repository provides a convenience method for recursive delete anyway, and a workaround would be to simply issue multiple CLI deletes.

**Comment by joshmoore** [comment]
re `DoAll`: this brings us back to `obj` and formerly `tx` since in some cases you'd like to do:
  * `chgrp`
  * `new Dataset`
  * `re-link`
  * other actions

re `dry-run`: sorry, @ximenesuk, I missed this. I was thinking this could list all the objects that would be processed rather than:
```
[jamoore@ome-c6100-3 ~]$ omero -q delete --dry-run Image:1578 --report
omero.cmd.Delete2 Image 1578... ok
Steps: 1
Elapsed time: 0.436 secs.
Flags: []
```

**Comment by ximenesuk** [comment]
Is it possible for the server-side `dryRun` to provide this information @mtbc ?

File: components/tools/OmeroPy/src/omero/plugins/group.py [hunk 1]
```diff
@@ -186,7 +185,9 @@ def perms(self, args):
                          % (g.name.val, gid, perms))
         else:
             try:
-                a.changePermissions(Grp(gid, False), perms)
+                chmod = omero.cmd.Chmod(
+                    type="/ExperimenterGroup", id=gid, permissions=str(perms))
+                c.sf.submit(chmod)
```
Comment (sbesson) [inline components/tools/OmeroPy/src/omero/plugins/group.py]: With this semantic change, the chmod command is submitted but returns immediately ending up in a race condition for the tests in https://ci.openmicroscopy.org/job/OMERO-5.1-merge-integration-python/424/testReport/test.integration.clitest.test_group/TestGroupRoot/testPerms___perms_rwra___rwr______id_/. As suggested by @joshmoore, we might want to use `c.submit()` instead.

**Comment by ximenesuk** [comment]
Fix pushed. Thanks @sbesson & @joshmoore (this also answers a query I had on another branch!)

**Comment by mtbc** [comment]
Yes, regardless of `dryRun` the response lists all the objects that would be processed and (separately for things like chgrp) also those that would be deleted (e.g., broken link).

**Comment by ximenesuk** [comment]
Thanks @mtbc, so the report option can be expanded for both cases. I'll look in to this.

**Comment by ximenesuk** [comment]
The one thing I have not dealt with here are graphspecs of the form `/Object+Only`  Should these simply map to `Object`? Or should an exception be thrown?

**Comment by joshmoore** [comment]
I would think they should throw an exception with a suggestion of what options would provide the same support. e.g. `/Image+Only` --> `Image --exclude...`

**Comment by ximenesuk** [comment]
@joshmoore happy to do that but I don't know what options would provide the same support! In the `GraphSpec` there are the following:
```
/Dataset+Only
/FileAnnotation+special
/Fileset+Only
/Image+Only
/Plate+Only
/Well+WithSamples
```

**Comment by ximenesuk** [comment]
The two failing tests are fixed but this raises more questions, detailed in the comments above.

More trivially the tests should be modified to use the argument form `Object:id` rather than `\Object:id` just to make the current usage clear.

**Comment by mtbc** [comment]
Can certainly chat today once you're here. I wasn't quite sure how/why some of the previous DoAll stuff worked. (!)

**Comment by ximenesuk** [comment]
@mtbc me neither! We can certainly have a chat. @joshmoore opinions on whether the two calls should have the same effect?

**Comment by joshmoore** [comment]
Sorry, I've lost the thread: which 2?

**Comment by ximenesuk** [comment]
The test in:

https://github.com/ximenesuk/openmicroscopy/commit/bbcf1f573de2ccf559efd69aabc1cf5f57f2c348

and the test at: 

https://github.com/ximenesuk/openmicroscopy/commit/bbcf1f573de2ccf559efd69aabc1cf5f57f2c348#diff-9b2f5789b1ddafb92cb7c1509d307f1cR236

have different outcomes.

**Comment by joshmoore** [comment]
And what was the status of the `DoAll` with 2 `Chgrp1` commands? My understanding would certainly be that 2 Chgrps in one DoAll transaction should allow this to happen, but this was previously implemented by special post-processing, and I don't know the status of that code now.

**Comment by mtbc** [comment]
Ah, that's true, the special processing muddies the water more. What happens now is special processing server-side to bundle the two Chgrp1 into one Chgrp2. (The thing I was confused about was apart from the processing.)

**Comment by ximenesuk** [comment]
With the two `Chgrp1` commands in a `DoAll` the test would have passed. This is effectively what was happening in the tests fixed in https://github.com/ximenesuk/openmicroscopy/commit/b0180f04483292f776376ffb1fd4ec319d5f6af2

**Comment by ximenesuk** [comment]
Following a conversation with @joshmoore and @mtbc the arguments should be combined into as few commands as possible unless an `--ordered` flag is used in which case they will be passed through strictly as they are listed.

**Comment by ximenesuk** [comment]
These later commits break some CLI tests. I'll be fixing those and adding new tests to check the command combining in a later commit.

**Comment by joshmoore** [comment]
@hflynn : this is going to take some *creative* release announcement text.

**Comment by ximenesuk** [comment]
It was the omission of the fix in the final commit that broke the tests. All CLI tests are now passing locally.

**Comment by ximenesuk** [comment]
@joshmoore "omero chgrp and delete commands improved to take advantage of new server-side functionality"?

**Comment by joshmoore** [comment]
@ximenesuk : no objection, but we've conceivably broken usage in a point release (because that previous usage was likely subtly broken...)

**Comment by ximenesuk** [comment]
These test do some basic checks for good cases where the combined and `--ordered` cases have the same outcome. There could probably me more tests and in particular tests for problem cases.

@sbesson I have resisted refactoring these classes at this stage, though I have made the odd correction. Some of the tests I have created could benefit from helpers not available on this branch.

**Comment by ximenesuk** [comment]
@pwalczysko I have modified the description, please let me know if this helps you understand the changes without needing to parse the comments and commits. Also, I hope this will help with testing and scenarios. Any questions just shout.

**Comment by pwalczysko** [comment]
```
[petr@LS25165 ~/Downloads/OMERO.server-5.1.0-517-02889da-ice35-b484]$ bin/omero delete /Image:2903 /Image:2904 /Image:2902 /Dataset:2403 /Dataset:2402
Previous session expired for user-2 on localhost:4064
Server: [localhost:4064]
Username: [user-2]user-6
Password:
Created session a3ee7148-fde7-4449-84a5-d98706418cb8 (user-6@localhost:4064). Idle timeout: 10.0 min. Current group: read-only-1
omero.cmd.Delete2 Image 2903,2904,2902, Dataset 2403,2402... ok
[petr@LS25165 ~/Downloads/OMERO.server-5.1.0-517-02889da-ice35-b484]$ 
```

The action passed as expected, but the message is not very clean: 
   - ``omer.cmd.Delete2`` somehow introduces info I do not want to know (that the ``Delete2`` method was used instead of Delete(1?) - just confusing, the ``2`` is not really reassuring, instead I would expected something like  ``.....were deleted successfully``

**Comment by joshmoore** [comment]
Rename works as excepted though I have questions about `templatePrefix` (filed to a card):
```
Renaming Fileset:3737 to user-4_5/2015-04/24/15-06-14.378/
Moving user-4_5/2015-04/24/14-07-51.434/ to user-4_5/2015-04/24/15-06-14.378/
Moving user-4_5/2015-04/24/14-07-51.434.log to user-4_5/2015-04/24/15-06-14.378.log
...
[hudson@ome-c6100-3 jamoore]$ ls /repositories/OMERO-5.1-merge-deploy/ManagedRepository/user-4_5/2015-04/24/14-07-51.434/
ls: cannot access /repositories/OMERO-5.1-merge-deploy/ManagedRepository/user-4_5/2015-04/24/14-07-51.434/: No such file or directory
[hudson@ome-c6100-3 jamoore]$ ls /repositories/OMERO-5.1-merge-deploy/ManagedRepository/user-4_5/2015-04/24/15-06-14.378
8kx8k.jpg
```
Logs as well:
```
[hudson@ome-c6100-3 jamoore]$ ls /repositories/OMERO-5.1-merge-deploy/ManagedRepository/user-4_5/2015-04/24/14-07-51.434.log
ls: cannot access /repositories/OMERO-5.1-merge-deploy/ManagedRepository/user-4_5/2015-04/24/14-07-51.434.log: No such file or directory
[hudson@ome-c6100-3 jamoore]$ ls /repositories/OMERO-5.1-merge-deploy/ManagedRepository/user-4_5/2015-04/24/15-06-14.378.log
/repositories/OMERO-5.1-merge-deploy/ManagedRepository/user-4_5/2015-04/24/15-06-14.378.log
```

**Comment by ximenesuk** [comment]
@pwalczysko the plugin output essentially echoes the type of the command object that is used (blame @joshmoore!). I think as developers we are used to a certain amount of terseness from cli output (beyond OMERO). However, it would be possible to provide something like you are asking for. Maybe, though, it should be an RFE for 5.1.2

**Comment by joshmoore** [comment]
`omero group perms` works as before (the change is admittedly much smaller than the others here):
```
[hudson@ome-c6100-3 jamoore]$ omero group perms --name 3725-chgrp-test --type read-annotate
Using session 77a85c56-8032-4da5-be78-b3461198ee33 (root@localhost:4064). Idle timeout: 10.0 min. Current group: system
Changed permissions for group 3725-chgrp-test (id=1203) to rwra--
[hudson@ome-c6100-3 jamoore]$ omero group perms --name 3725-chgrp-test --perms rw----
Using session 77a85c56-8032-4da5-be78-b3461198ee33 (root@localhost:4064). Idle timeout: 10.0 min. Current group: system
Changed permissions for group 3725-chgrp-test (id=1203) to rw----
[hudson@ome-c6100-3 jamoore]$ omero group perms --name 3725-chgrp-test --perms rwrwrw
Using session 77a85c56-8032-4da5-be78-b3461198ee33 (root@localhost:4064). Idle timeout: 10.0 min. Current group: system
Changed permissions for group 3725-chgrp-test (id=1203) to rwrwrw
[hudson@ome-c6100-3 jamoore]$ omero group perms --name 3725-chgrp-test --perms rw----
Using session 77a85c56-8032-4da5-be78-b3461198ee33 (root@localhost:4064). Idle timeout: 10.0 min. Current group: system
Changed permissions for group 3725-chgrp-test (id=1203) to rw----
```

**Comment by joshmoore** [comment]
MIF delete looks good:
```
[hudson@ome-c6100-3 tmp]$ omero delete Image:253493
omero.cmd.Delete2 Image 253493... failed: 'graph-fail'
failed: may not split Fileset[3744]
[hudson@ome-c6100-3 tmp]$ omero delete Image:253494
omero.cmd.Delete2 Image 253494... failed: 'graph-fail'
failed: may not split Fileset[3744]
[hudson@ome-c6100-3 tmp]$ omero delete Image:253493 Image:2534934 --ordered
omero.cmd.Delete2 Image 253493, Image 2534934... failed: 'graph-fail'
failed: may not split Fileset[3744]
[hudson@ome-c6100-3 tmp]$ omero delete Image:253493 Image:253494
omero.cmd.Delete2 Image 253493,253494... ok
```

**Comment by pwalczysko** [comment]
Delete of multiple objects works with the old slash syntax as well as the new syntax.
Delete of one object with --exclude Annotations works as well when one annotation and exclusive for the object.

**Comment by pwalczysko** [comment]
Bug
select multiple images, annotate all of them with one tag. execute following sequence of commands.


```
[petr@LS25165 ~/Downloads/OMERO.server-5.1.0-517-02889da-ice35-b484]$ bin/omero delete --exclude Annotation Image:6217
Previous session expired for user-6 on localhost:4064
Server: [localhost:4064]
Username: [user-6]ome
Password:
Password check failed for 'ome': []
Password:Cancelled
InternalException: Failed to connect: die called: Cancelled
[petr@LS25165 ~/Downloads/OMERO.server-5.1.0-517-02889da-ice35-b484]$ bin/omero delete --exclude Annotation Image:6217
Previously logged in to localhost:4064 as user-6
Server: [localhost:4064]
Username: [user-6]
Password:
Created session e80825fb-d2a7-4041-8da2-8e317d0de5f4 (user-6@localhost:4064). Idle timeout: 10.0 min. Current group: read-only-1
omero.cmd.Delete2 Image 6217... ok
[petr@LS25165 ~/Downloads/OMERO.server-5.1.0-517-02889da-ice35-b484]$ bin/omero delete Image:6221
Using session e80825fb-d2a7-4041-8da2-8e317d0de5f4 (user-6@localhost:4064). Idle timeout: 10.0 min. Current group: read-only-1
omero.cmd.Delete2 Image 6221... ok
[petr@LS25165 ~/Downloads/OMERO.server-5.1.0-517-02889da-ice35-b484]$ bin/omero delete Image:6213,6214,6211 --exclude Annotations
Using session e80825fb-d2a7-4041-8da2-8e317d0de5f4 (user-6@localhost:4064). Idle timeout: 10.0 min. Current group: read-only-1
omero.cmd.Delete2 Image 6213,6214,6211... failed: 'steps-cancel'
	message=
	stacktrace=java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:213)
	at com.google.common.collect.SingletonImmutableSet.<init>(SingletonImmutableSet.java:47)
	at com.google.common.collect.ImmutableSet.of(ImmutableSet.java:94)
	at com.google.common.collect.ImmutableSet.construct(ImmutableSet.java:185)
	at com.google.common.collect.ImmutableSet.copyOf(ImmutableSet.java:375)
	at omero.cmd.graphs.ChildOptionI.init(ChildOptionI.java:109)
	at omero.cmd.graphs.Delete2I.init(Delete2I.java:115)
	at omero.cmd.HandleI.steps(HandleI.java:427)
	at omero.cmd.HandleI$1.doWork(HandleI.java:365)
	at omero.cmd.HandleI$1.doWork(HandleI.java:361)
	at sun.reflect.GeneratedMethodAccessor299.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:566)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy72.doWork(Unknown Source)
	at ome.services.util.Executor$Impl.execute(Executor.java:447)
	at omero.cmd.HandleI.run(HandleI.java:359)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at ome.services.util.Executor$Impl$1.call(Executor.java:488)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)


[petr@LS25165 ~/Downloads/OMERO.server-5.1.0-517-02889da-ice35-b484]$ 
```

**Comment by joshmoore** [comment]
> [petr@LS25165 ~/Downloads/OMERO.server-5.1.0-517-02889da-ice35-b484]$ bin/omero delete Image:6213,6214,6211 --exclude Annotations

Perhaps a typo? Just `--exclude Annotation`

**Comment by joshmoore** [comment]
NB: for `--dry-run --report` I'd suggest only having parseable output on stdout:
```
Deleted ome.model.display.QuantumDef objects
ome.model.display.QuantumDef:28393
ome.model.display.QuantumDef:28394
```
becomes
```
ome.model.display.QuantumDef:28393
ome.model.display.QuantumDef:28394
```

**Comment by mtbc** [comment]
Yes, needs to be `Annotation` singular. Can also name specific types of annotation; the names on http://www.openmicroscopy.org/site/support/omero5.1/developers/Model/EveryObject.html ought to work.

**Comment by mtbc** [comment]
At some point I'm going to have to improve some of these error messages, though. Oh, for a sufficiently rainy day ...

**Comment by pwalczysko** [comment]
Yes, Annotation singular worked fine.

**Comment by joshmoore** [comment]
@ximenesuk & @mtbc, thanks for all the heroics on getting this done. There are no known issues with this, but talking to @pwalczysko EOB today, it seems this is still the scariest bit left dangling. The exact words were "quite a chunk of work". I'm adding the `exclude` label, and we can get this in during the early days of 5.1.2.

**Comment by ximenesuk** [comment]
@joshmoore I think that makes sense. I can then address some of the minor issues and documentation too as well as developing some more robust tests that keep @pwalczysko happy. Thanks all.

**Comment by ximenesuk** [comment]
Sorry, post-pub ramblings but given what @pwalczysko  has commented above I think the CLI default action should be that of the other clients if possible. So if when deleting an image you are asked if you also want to delete the annotations then similarly here,
```
omero delete Image:101
```
should **not** delete the annotations by default as it does now but
```
omero delete Image:101 --include Annotation
```
should be required to delete also annotations. Similarly for any other delete/chgrp via the GUI clients. This would not only make the command safer, especially for users, but also have all the clients do the same thing.

/cc @joshmoore @mtbc

**Comment by ximenesuk** [comment]
@joshmoore re https://github.com/openmicroscopy/openmicroscopy/pull/3725#issuecomment-95965525 yes I pretty much translated the existing report as is. I would go further and suggest a compressed report:
```
ome.model.display.QuantumDef:28393,28394
```
at least as an option.

**Comment by joshmoore** [comment]
* :+1: default of `--exclude Annotation`
 * :+1: on compact format. (Might want to drop `ome.model.display` though)

**Comment by ximenesuk** [comment]
@joshmoore I think both of these could be done client-side but I don't know if they would benefit from server-side changes and if that were the case how that would affect other clients. /cc @mtbc

**Comment by mtbc** [comment]
I wonder what clients actually send if you do or don't click that checkbox. `--exclude Annotation` will even leave comments and ratings remaining after the image is deleted.

**Comment by ximenesuk** [comment]
@mtbc I'll take a look at the client code since mirroring what they do would make sense. I don't know if any other clients consume the extended object-by-object output though so maybe that would be something that could modified easily server-side if there was a map of what was and wasn't useful.

**Comment by ximenesuk** [comment]
Does this PR need any of the comments/RFEs mentioned above addressing before this can be considered for merging or does it just need further review? I would prefer to merge this as early as possible and then add a card for the various RFEs.

**Comment by joshmoore** [comment]
With 5.1.1 behind us, either way works for me.

**Comment by sbesson** [comment]
To allow further testing/improvements/subsequent work by various contributors, merging this now 5.1.1 is out. Thanks for the CLI overhaul @ximenesuk.
